gum/gumscriptcore.cpp

```diff
@@ -855,112 +861,40 @@ gum_script_core_on_native_pointer_is_null (
-  }
-  else
-  {
-    rhs = info[0]->ToInteger ()->Value ();
-  }
-  gpointer result = GSIZE_TO_POINTER (lhs - rhs);
-
-  info.GetReturnValue ().Set (_gum_script_pointer_new (result, self));
-}
-
-static void
-gum_script_core_on_native_pointer_and (
-    const FunctionCallbackInfo<Value> & info)
-{
-  GumScriptCore * self = static_cast<GumScriptCore *> (
-      info.Data ().As<External> ()->Value ());
-
-  guint64 lhs = reinterpret_cast<guint64> (
-      GUM_NATIVE_POINTER_VALUE (info.Holder ()));
-
-  guint64 rhs;
-  Local<FunctionTemplate> native_pointer (
-    Local<FunctionTemplate>::New (self->isolate, *self->native_pointer));
-  if (native_pointer->HasInstance (info[0]))
-  {
-    rhs = reinterpret_cast<guint64> (
-        GUM_NATIVE_POINTER_VALUE (info[0].As<Object> ()));
-  }
-  else
-  {
-    rhs = info[0]->ToInteger ()->Value ();
-  }
-  gpointer result = GSIZE_TO_POINTER (lhs & rhs);
-
-  info.GetReturnValue ().Set (_gum_script_pointer_new (result, self));
-}
-
-static void
-gum_script_core_on_native_pointer_or (
-    const FunctionCallbackInfo<Value> & info)
-{
-  GumScriptCore * self = static_cast<GumScriptCore *> (
-      info.Data ().As<External> ()->Value ());
-
-  guint64 lhs = reinterpret_cast<guint64> (
-      GUM_NATIVE_POINTER_VALUE (info.Holder ()));
-
-  guint64 rhs;
-  Local<FunctionTemplate> native_pointer (
-    Local<FunctionTemplate>::New (self->isolate, *self->native_pointer));
-  if (native_pointer->HasInstance (info[0]))
-  {
-    rhs = reinterpret_cast<guint64> (
-        GUM_NATIVE_POINTER_VALUE (info[0].As<Object> ()));
-  }
-  else
-  {
-    rhs = info[0]->ToInteger ()->Value ();
-  }
-  gpointer result = GSIZE_TO_POINTER (lhs | rhs);
+#define NATIVE_POINTER_OP_IMP(name, op) \
+    static void \
+    gum_script_core_on_native_pointer_ ## name ( \
+        const FunctionCallbackInfo<Value> & info) \
+    { \
+        GumScriptCore * self = static_cast<GumScriptCore *> ( \
+            info.Data ().As<External> ()->Value ()); \
+        \
+        guint64 lhs = reinterpret_cast<guint64> ( \
+            GUM_NATIVE_POINTER_VALUE (info.Holder ())); \
+        \
+        guint64 rhs; \
+        Local<FunctionTemplate> native_pointer ( \
+            Local<FunctionTemplate>::New (self->isolate, \
+                *self->native_pointer)); \
+        if (native_pointer->HasInstance (info[0])) \
+        { \
+          rhs = reinterpret_cast<guint64> ( \
+              GUM_NATIVE_POINTER_VALUE (info[0].As<Object> ())); \
+        } \
+        else \
+        { \
+          rhs = info[0]->ToInteger ()->Value (); \
+        } \
+        gpointer result = GSIZE_TO_POINTER (lhs op rhs); \
+        \
+        info.GetReturnValue ().Set (_gum_script_pointer_new (result, self)); \
+    }
 
-  info.GetReturnValue ().Set (_gum_script_pointer_new (result, self));
-}
+NATIVE_POINTER_OP_IMP (add, +)
+NATIVE_POINTER_OP_IMP (sub, -)
+NATIVE_POINTER_OP_IMP (and, &)
+NATIVE_POINTER_OP_IMP (or,  |)
+NATIVE_POINTER_OP_IMP (xor, ^)
```

Comment:
Would be great to (sort of) follow the convention we have here:
https://github.com/frida/frida-gum/blob/master/gum/gumscriptmemory.cpp#L131
So something like `GUM_DEFINE_NATIVE_POINTER_OP_IMPL`.